charms/istio-pilot/src/istio_gateway_name_provider.py

```diff
@@ -0,0 +1,25 @@
+import logging
+from ops.framework import Object, EventBase, EventSource, ObjectEvents
+from ops.charm import CharmEvents
+from lightkube.core.client import Client
+from lightkube.generic_resource import create_namespaced_resource
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_RELATION_NAME = "gateway"
+
+
+class GatewayProvider(Object):
+    def __init__(self, charm, relation_name=DEFAULT_RELATION_NAME):
+        super().__init__(charm, relation_name)
+
+    def on_gateway_relation_changed(self, charm, gateway_name, gateway_namespace):
+        if self.model.unit.is_leader():
```

Comment:
I don't know if I prefer the is_leader check here or in the charm itself.  If you haven't thought about it since moving this out of a library, give it some though and go with whichever you prefer.

Regardless of where it is, we should do some logging if not leader.  As is, I could see it being really hard to debug if the charm does `gateway_object.on_gateway_relation_changed(...)` and sees no exceptions, no logging, and no change in the relation data.  I'm not sure if this should raise an exception that is caught by the charm, or just log that no data was sent.  But something should happen